Add resources, sources, and references during create component flow

jmickey opened the pull request:
Enable users to add resources, sources, and references within the Create Component UI flow.

- [x] Add resources (file, directory, ociImage)
- [x] Add sources
- [x] Add references

Changes:

- Added buttons for "Add Resource", "Add Source", and "Add Reference"
- Adding a resources allows the user to specify a resource type, current supported types are "file", "directory", and "ociImage".
- Adding a source currently support a gitHub source.
- Adding a reference supports a basic reference specification of `name`, `version`, and `componentName`. Support for specifying a digest can be completed later.
- If an error occurs adding any of the above the process exits early and the user will need to rectify.

Current Limitations:

- Once you add a resource, source, or reference, you can't remove them. This requires some additional considerations.
- Only basic input validation is completed, which only validates that the fields are not empty.

phoban01 commented on src/commands/addResources.ts:
I think using the a resource file would be a more straightforward way to approach this, We could just serialize the resource, create a temporary file and pass it to the CLI

phoban01 commented on src/commands/addSource.ts:
Ditto previous comment re input file

phoban01 commented on src/webviews/createComponent.ts:
```diff
@@ -21,16 +22,67 @@ interface CreateComponentWebviewContent {
   };
 }
 
+export interface FileResource {
```
We should use the ocm types already defined here: https://github.com/open-component-model/vscode-ocm-tools/blob/main/src/ocm/ocmv3.d.ts

phoban01 reviewed (changes requested):
Approach looks good. I think it can be improved by using the input file and leveraging the generated OCM types.

phoban01 commented:
It would be nice to be able to specify a path within the workspace in which to create the component.

phoban01 commented:
For some reason when a component is created using the extension it is not possible to remove it from the treeview, even if it remove the `component-archive directory.

I have validated this using the following process:
- create a component using the CLI 
- refresh "Local Components" treeview
- component is listed in treeview
- delete the component archive directory
- refresh "Local Components" treeview
- component  is removed from treeview
- create component using the extension
- refresh "Local Components" treeview
- component is listed in treeview
- delete the component archive directory
- refresh "Local Components" treeview
- component remains listed in treeview (when it should have been removed)

jmickey commented:
@phoban01 

> Adding a Source currently doesn't seem to work.

This is resolved.

> For some reason when a component is created using the extension it is not possible to remove it from the treeview

I noticed this too, and tracking down why was pretty difficult, but I believe I have fixed the issue in the latest commit. Refresh behaviour is a little opaque, but I found this nice example from the MS team: https://github.com/microsoft/vscode-extension-samples/blob/main/tree-view-sample/src/nodeDependencies.ts, which helped me resolve this. 

Essentially, the tree wasn't correctly being rebuilt when refresh was called, which in the example is being done in the `getChildren` method. This lead me to notice that the internal `treeViews` state isn't really required, so I removed it. From my manual testing the refresh functionality is now working as intended with the current changes.

> It would be nice to be able to specify a path within the workspace in which to create the component.

Should be easy enough.

phoban01 commented:
Verified that Sources can be added and that refreshing works as expected now 🥇

phoban01 commented on src/commands/createComponent.ts:
```diff
@@ -32,21 +42,73 @@ export async function createComponent(
+    for (const resource of component.resources) {
+      const res: AddResourceResult = await addResource(resource, outFile);
+      if (!res.success) {
+        // return early if a resource could not be added.
+        workspaceTreeViewProvider.refresh();
+        return {
+          success: false,
+          message: `Component creation failed adding resource: ${res.message}`,
+        };
+      }
+    }
+  }
+
+  if (component.references && component.references.length > 0) {
+    for (const reference of component.references) {
+      const res: AddReferenceResult = await addReference(reference, outFile);
+      if (!res.success) {
+        // return early if a reference could not be added.
+        workspaceTreeViewProvider.refresh();
+        return {
+          success: false,
+          message: `Component creation failed adding reference: ${res.message}`,
+        };
+      }
+    }
+  }
+
+  if (component.sources && component.sources.length > 0) {
+    for (const source of component.sources) {
+      const res: AddSourceResult = await addSource(source, outFile);
+      if (!res.success) {
+        // return early if a source could not be added.
+        workspaceTreeViewProvider.refresh();
+        return {
+          success: false,
+          message: `Component creation failed adding source: ${res.message}`,
+        };
+      }
+    }
+  }
```
this is nice 👍🏻

phoban01 reviewed (approved):
Looks great, thanks Josh!
